quixstreams/sources/community/file/file.py

```diff
@@ -16,6 +19,54 @@
 logger = logging.getLogger(__name__)
 
 
+class FileFetcher:
+    """
+    Serves individual files while downloading another in the background.
+    """
+
+    def __init__(self, origin: Origin, filepath: Path):
+        self._origin = origin
+        self._file_names = iter(self._origin.file_collector(filepath))
+        self._stopped: bool = False
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._downloading_file_name: Optional[Path] = None
+        self._downloading_file_content: Optional[Future] = None
+        self._download_next_file()
+
+    def __iter__(self) -> Self:
+        return self
+
+    def __next__(self) -> tuple[Path, BinaryIO]:
+        if self._stopped:
+            raise StopIteration
+
+        try:
+            file_name = self._downloading_file_name
+            file_content = self._downloading_file_content.result()
+            self._download_next_file()
+            return file_name, file_content
+        except Exception as e:
+            logger.error("FileFetcher encountered an error", exc_info=e)
+            self.stop()
+            raise e
+
+    def stop(self):
+        logger.info("Stopping file download thread...")
+        self._stopped = True
+        self._executor.shutdown(wait=False)
```

Comment:
we should cancel the in-progress future before shutting down.

We also should `wait=True` the shutdown to make it clean

Reply:
So the original reason I did this was to hopefully kill the file download immediately rather than waiting to complete it, as it seems you cant in actuality cancel already running futures (you can still call cancel on it, but from what I can tell, it won't do anything). 

However, I think for now I'll do a wait, and later on if we really want to we can add a mid-download stop (where you read chunks of data of the file at a time and look for stopping). I can't imagine people will be stopping in the middle very often anyway.

Reply:
In any case we have the source shutdown timeout